Cookbook/Proposals/MealyReactiveFeedback.md

```diff
@@ -0,0 +1,273 @@
+
+#### Example 2: Navigation
+
+```swift
+let externalEvent = Signal.merge(
+    didTapA.map { Event.route(Route.pushA) },
+    didTapB.map { Event.route(Route.modalB) }
+)
+
+let navigationFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        // NOTE: Using `flatMap(.first)` prevents from routing to multiple screens simultaneously
+        .flatMap(.first) { event, _ in
+            guard case let route = event.route else { return .empty }
+            return flowController.handle(route)
+        }
+        .ignoreValues() // no feedback loop
+        .promoteValue() // for adjusting type
+}
+
+let system = Property(
+    initial: 0,
+    reduce: reducer,
+    events: externalEvent,
+    feedbacks: [navigationFeedback]
+)
+```
+
+For navigation feedback rule and `flatMap(.first)`, see also [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44).
+This is a handy FRP way of managing "isTransitioning" state outside of `State`.
+
+Because this routing is now event-driven rather than state-driven, **we don't need to transit the state back to the original position after completed, which will reduce the intermediate `case`s e.g. `State.isPushingA` and `Event.didFinishPushingA`**.
+(But note that we might still want to have these cases if we want to precisely handle tem to avoid interfering with other events/states/feedbacks)
+
+#### Example 3: Image Loading
+
+```swift
+enum Event {
+    case callAPI(Request)
+    case fetchImages(Response)
+    case reloadCell(image: UIImage, index: Int)
+}
+
+struct State {
+    var images: [/* cellIndex */ Int: UIImage?] = [:]
+}
+
+let externalEvent = Signal.merge(
+    viewDidLoad.map { Event.callAPI(Request(apiURL)) }
+}
+
+let apiFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        .flatMap(.first) { event, _ in
+            guard case let .callAPI(request) = event else { return .empty }
+            return apiSession.send(request)
+        }
+        .map(Event.fetchImages)
+}
+
+let imageFeedback = Feedback { signal -> Signal<Event> in
+    return signal
+        .flatMap(.merge) { event, _ -> Signal<(UIImage?, index: Int)> in
+            guard case let .fetchImages(response) = event else { return .empty }
+            return imageLoader.fetchAllImages(response.imageURLs)
+        }
+        .map { image, index in
+            Event.reloadCell(image: image, index: index)
+        }
+}
+
+let reducer = { state, event in
+    guard case let .reloadCell(image, index) = event else { return state }
+
+    var state = state
+    state.images[index] = image
+    return state
+}
+
+let system = Property(
+    initial: State(),
+    reduce: reducer,
+    events: externalEvent,
+    feedbacks: [apiFeedback, imageFeedback]
+)
+
+system.signal
+    .combinePrevious(initial: system.value)
+    .startWithValues { oldState, newState in
+        // NOTE: diff & patch algorithm must be efficient enough to update images only
```

Comment:
A remark on efficiency.

In our world of pure `virtual component_tree = f(state)`,  the virtual component tree would contain the image, and the diff-patching is done by Bento.

Assuming there is an arbitrarily large number of items, my educated assumption is that the diff-patching Bento is unlikely our most prominent bottleneck. The linear cost of the virtual component tree recomputation, against many state change occurrences cased by small updates, is more likely to first be the issue. Most importantly, this is inherently the cons of the pure model we have, and perfect diff-patching does not help alleviate this in any manner.

Reply:
It might also be better to refactor this example closer to what would actually happen in our codebase, if it would still be relevant afterwards to the topic at hand.

e.g. `Action.itemWillAppear` reports item appearance to VM, where the VM initiates the network request. Similarly, VM responds to `Action.itemDidDisappear` by cancelling requests or purging stuff as needed.

Reply:
Thanks for detailed explanation here.
I have to admit that this example is rather my thought experiment (for Bento without `Reactive.Property` issue) that might not work well in practice.
Though I believe handling view lifecycles e.g. `itemWillAppear` are still possible, managing every remote image rendering states may be boring and not really interested, as Elm or React are not even doing this kind of micro-management AFAIK.

I think I should remove this example.